### ISSUE TITLE
flaws in OpenSSL-1.1.1-dev builds on Windows

### ISSUE THREAD

**Issue by sav-ix** [openssl/openssl#3254]
Hello everyone,

There are few issues with OpenSSL-1.1.1-dev builds on Windows:
- [x] 1. [out of scope](https://github.com/openssl/openssl/issues/3254#issuecomment-296153751).

for Debug+Static build using MSVC with '`-MTd`' key got warnings:
```
perl Configure -MTd threads no-deprecated no-shared debug-VC-WIN64A && nmake
[snip]
cl : Command line warning D9025 : overriding '/MTd' with '/MT'
[snip]
cl : Command line warning D9025 : overriding '/MTd' with '/MDd'
```
After investigation of build logs with different configurations found, that OpenSSL build system add the following compiler keys:
- '`/MDd`' for Debug+Shared build,
- '`/MD`' for Release+Shared build,
- '`/MT`' for Release+Static build,
as expected according to its [description](https://msdn.microsoft.com/en-us/library/2kzt1wy3.aspx). But for some reason it add
- '`/MT`' and '`/MDd`' keys, instead of '`/MTd`', for Debug+Static build.

The only information on this topic which I found, is a comment:
```
# Following might/should appears controversial, i.e. defining
# /MDd without evaluating $disabled{shared}. It works in
# non-shared build because static library is compiled with /Zl
# and bares no reference to specific RTL. And it works in
# shared build because multiple /MDd options are not prohibited.
# But why /MDd in static build? Well, basically this is just a
# reference point, which allows to catch eventual errors that
# would prevent those who want to wrap OpenSSL into own .DLL.
# Why not /MD in release build then? Well, some are likely to
# prefer [non-debug] openssl.exe to be free from Micorosoft RTL
# redistributable.
```
in file '`Configurations/10-main.conf`'. Though it doesn't mention '`/MT`' key, it obviously relate to this issue.

Not sure, why it was implemented just for Debug+Static builds. Is it possible to set compiler key '`/MTd`' as default for OpenSSL Debug+Static builds using MSVC? (or at least in places, where '`/MT`' key was set) For consistency purpose and in order to get the same build options with other libraries, which use OpenSSL (e.g. Qt).
CC:@dot-asm,@levitte

- [x] 2. [out of scope](https://github.com/openssl/openssl/issues/3254#issuecomment-390409527).

'`cflags`' variable from '`VC-common`' toolchain in file '`Configurations/10-main.conf`' contain compiler keys '`-Gs0 -GF -Gy`' ('`/Gs0`' == '[`/Gs`](https://msdn.microsoft.com/en-us/library/9598wk25.aspx)'), which already applied for MSVC Release builds via '[`/O2`](https://msdn.microsoft.com/en-us/library/8f8h5cxt.aspx)' key.

Therefore explicitly setting these keys affects MSVC Debug builds only. In such case doesn't it make sense to move them to '`/Od`' key in '`VC-noCE-common`' toolchain to emphasize, that they are intended for Debug builds only? (or drop them, if they aren't required for MSVC Debug builds)

- [x] 3. [unix-Makefile.tmpl](https://github.com/openssl/openssl/blob/master/Configurations/unix-Makefile.tmpl) file contains `-std=c++11` flag. Does it mean OpenSSL require C++11 compiler, and if so, why it's not mentioned in documentation?

- [x] 4. for builds using MSVC with ActivePerl got error:
```
Use of uninitialized value $platform in concatenation (.) or string at util\mkbuildinf.pl line 17.
```
not reproduced for builds using mingw-w64 with MSYS2 Perl.

- [x] 5. for builds using mingw-w64 and MSVC during '`make install`' got errors:
```
Cannot find "BIO_read_ex" in podpath: cannot find suitable replacement path, cannot resolve link
Cannot find "d2i_SCT_LIST" in podpath: cannot find suitable replacement path, cannot resolve link
Cannot find "d2i_SCT_LIST" in podpath: cannot find suitable replacement path, cannot resolve link
Cannot find "i2d_SCT_LIST" in podpath: cannot find suitable replacement path, cannot resolve link
Cannot find "CRYPTO_set_mem_functions" in podpath: cannot find suitable replacement path, cannot resolve link
Cannot find "SSL_bytes_to_cipher_list" in podpath: cannot find suitable replacement path, cannot resolve link
Cannot find "SSL_CTX_set0_ctlog_store" in podpath: cannot find suitable replacement path, cannot resolve link
```

- [x] 6. MSVC installation contain files:
```
include/openssl/__DECC_INCLUDE_EPILOGUE.H
include/openssl/__DECC_INCLUDE_PROLOGUE.H
```
which doesn't exist in mingw-w64 installation. Since they contain comment:
```
 * This file is only used by HP C on VMS, and is included automatically
 * after each header file from this directory
```
do they relate to MSVC builds?

- [ ] 7. mingw-w64 installation contain files:
```
ssl/misc/CA.pl
ssl/misc/tsget 
```
while MSVC contain files:
```
ssl/misc/CA.pl
ssl/misc/tsget.pl
```
Since files '`CA.pl`' has identical '`.pl`' extension in both installations, why files '`tsget(.pl)`' has different?

- [x] 8. [out of scope](https://github.com/openssl/openssl/issues/3254#issuecomment-390380684).

mingw-w64 builds print warnings into stderr (see [openssl1.1.1_mingw-w64_stderr.log](https://github.com/openssl/openssl/files/1974350/openssl1.1.1_mingw-w64_stderr.log)), while there are no such for MSVC builds.

- [x] 9. For builds with `zlib-dynamic` flag using mingw-w64 got errors:
```
11232:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(ZLIB1.dll)
6988:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(ZLIB1.dll)
5684:error:25078067:DSO support routines:win32_load:could not load the shared library:crypto/dso/dso_win32.c:107:filename(ZLIB1.dll)
```
during `make test` run. This is odd, since mingw-w64 provide embedded zlib, which expected to be used.
It should also be noted, that mingw-w64 installation folder doesn't contain any `zlib*.dll` files.

Environment:
- Windows 10 x64,
- MSVC 2017 15.5.0,
- Windows SDK 10.0.16299.15,
- Active Perl x64 5.24.3.2404,
- mingw-w64 x86_64 6.3.0,
- MSYS2 20170918 (with recent updates),
- OpenSSL-1.1.1-dev (bc624bd95554ea218cbc3fad0d794abdfeb21eb6).

Let me know, if anything of above worth to open a separate issue.

Best,

Alexander

**Comment by dot-asm** [comment]
> for Debug+Static build using MSVC with '-MTd' key got warnings:

If it means that you passed -MTd as additional argument to Configure, then just don't.

> /MT' and '/MDd' keys, instead of '/MTd', for Debug+Static build.

It's intentional, as discussed in referred comment. And formally it's not just /MT, but combination of /MT and /Zl. And in combination with /Zl it really doesn't matter if it's /MT or /MTd.

> For consistency purpose and in order to get the same build options with other libraries, which use OpenSSL (e.g. Qt).

If you build static OpenSSL library, you can use it in any way you like with another application or toolkit. In sense that you can use /Mx flag of *your choice* and link with OpenSSL static .libs. Or in other words, OpenSSL static .libs are agnostic to application compile flags, and in a sense auto-consistent with any external application. And controversial combination is kind of insurance that you can do that.

**Comment by levitte** [comment]
@sav-ix, could you check if this issue is resolved by now?  There's been a number of changes in the configuration stuff since last year...

**Comment by sav-ix** [comment]
Done.

**Comment by levitte** [comment]
So, uhmmm, not sure I understand.  You've "checked" some of the items in your list but not others.  Does that mean that those that are "checked" have been resolved, or are those the ones we still need to look into?

**Comment by sav-ix** [comment]
Checked items are resolved ones.
Made in accordance with to GitHub [Markdown](https://guides.github.com/features/mastering-markdown/)
![github_markdown](https://user-images.githubusercontent.com/3617101/39671622-588e9de4-5124-11e8-8fd1-e15a406ea4e9.png)

**Comment by levitte** [comment]
Thanks.  I know the mechanics for doing what you've done, but I needed to know what you meant with it (mind reading isn't one of my skills).

**Comment by levitte** [comment]
2. Not sure what you want here.  There's a difference between CFLAGS (user override is supported) and cflags (we want these flags unconditionally)

5. Those are not errors, they are warnings, and fairly stupid ones to boot.  What that says is that the processor has found a link to `BIO_read_ex` (`L<BIO_read_ex>`) but cannot find the corresponding .pod file (because that particular function is documented in `BIO_read.pod`, and the issue is resolved with symlinking).  So just ignore those lines.

6. I actually haven't thought about where those files are installed or not, we have not made anything special to avoid them being installed everywhere.  However, under a POSIX shell such as mingw, file name casing matters, while in `cmd`, it doesn't, so copying `*.h` gives different results depending on the environment.  WONTFIX for the moment.

7. Windows can associate extensions to interpreters, while a POSIX shell doesn't need that...  however, you're probably right, considering that stuff built with a mingw target may be used outside of the POSIX shell, so we should probably keep the extension there as well.  That would apply to Cygwin builds as well, right?

8. I'm not sure what you mean...  but I do know from experience that `cl` writes certain things to stdout where it would make more sense to me if they were written to stderr.  I currently don't remember where it prints errors and warnings, but I suspect it's stdout.  Or are you saying that mingw gives some warnings that cl doesn't warn about at all?  Let's remember that we're talking about different compilers and that the mingw build attempts to look like POSIX while MSVC doesn't, at all.

**Comment by levitte** [comment]
(meh, markdown gave me some trouble...  comment above now edited too fit the original numbering )

**Comment by levitte** [comment]
2 I would be OK with that.  @dot-asm, what say you?

5 I just dicovered that the `--quiet` flag silences these kinds of warnings only.  I'm usually a bit hesitant to turn off all warnings, but in this case, it seems appropriate.  See #6304

6 See #6303

7 I think that considering expectations on different platforms, I'm gonna leave this alone, to be revisited in the next major release.

8 Ok, so compiler differences.  Not an issue.

So in summary, all but 2 have been covered by now.

**Comment by dot-asm** [comment]
> 2 I would be OK with that. @dot-asm, what say you?

I can't confirm assertion that /Gs0 is part of /O2. There is some /Gs limit implied in /O2(*), but it's non-zero. What does /GsN mean? In means "generate calls to __chkstk", and the number controls "when", with 0 meaning "always". With last part in mind, i.e. /Gs0 *always* generating the call, one can wonder if it's actually appropriate or even meant to be that way. It does add some overhead and hence impacts efficiency of short functions. It might happen that it was added last century for some reason without giving it deeper consideration and was never looked back upon. In other words it's arguably appropriate to make /Gs0 debug-only flag. But not because it's part of /O2, but because /Gs0 is crippling the performance. As for remaining /GF /Gy. Well, it's rather matter of taste, but I for one would argue for keeping them where they are now. It's a way of being explicit about what's common for both production and debug builds. If they are moved to debug flags, then you'd have to keep in mind that they are part of /O2. And you're going to forget, ask yourself what-the-heck, dig in documentation, say a-a-a-a-h!, then forget, what-the-heck, etc... Besides, note that /O2 is in CFLAGS, i.e. intended to be user-overrideable, while we do want to enforce at least /GF. Because we would rather see our code crash if it tries to write to const string irregardless build or user-supplied flags. (Well, it's not like user stands no chance of working around the obstacle, but at least we arranged in a way that tells "you're on your own if you do".)

(*) Actually one can wonder if some /Gs limit is always implied. Because it's also about how application interacts with OS. More specifically you're supposed to touch pages on stack in sequence, and __chkstk does it for you, so excessive stack allocations are better checked with __chkstk, and this applies irregardless of compiler flags.

**Comment by sav-ix** [comment]
From the [/Gs](https://msdn.microsoft.com/en-us/library/9598wk25.aspx):
> If the /Gs option is specified without a size argument, it is the same as specifying /Gs0.

and  [/O2](https://msdn.microsoft.com/en-us/library/8f8h5cxt.aspx):
> /O2 (Maximize Speed) Equivalent to /Og /Oi /Ot /Oy /Ob2 /Gs /GF /Gy

documentation follows, that /Gs0 is a part of /O2. If not, does:
> I can't confirm assertion that /Gs0 is part of /O2. There is some /Gs limit implied in /O2(*), but it's non-zero

mean, that there is inaccuracy in MSVC documentation?

As for `/GF /Gy` location, the only reason to move them to Debug flags, that they has no effect in Release flags. Also no need to keeping in mind the reasons of such move, since `git blame` can remind that in a second.
But if keeping them where they are now would improve code cleanliness and security, then it's better to refrain from such changes.

**Comment by dot-asm** [comment]
Take for example `void foo() { char a[16]; bar(a); }`, compile with /O2 and with /O2 /Gs and compare.

**Comment by sav-ix** [comment]
@dot-asm
> Take for example void foo() { char a[16]; bar(a); }, compile with /O2 and with /O2 /Gs and compare.

I confirm, `/O2` and `/O2 /Gs` provide different binary output. Yet another reason to leave MSVC flags as is.

@levitte
> I just dicovered that the --quiet flag silences these kinds of warnings only. I'm usually a bit hesitant to turn off all warnings, but in this case, it seems appropriate. See #6304

After PR# dd1477ced7f3679d8177f942219e96fa86b37234 was merged, all `Cannot find "..." in podpath ...` warnings gone.
Hope this solution won't hide important warnings, which can occur in future.


> Windows: don't install __DECC_*.H

12843c7307900e124ee0407535cab5534fc27dad didn't work for me. Can you reopen https://github.com/openssl/openssl/pull/6303 ?

**Comment by levitte** [comment]
> 12843c7 didn't work for me. Can you reopen #6303 ?

Sure, have a look at #6417

**Comment by levitte** [comment]
Never mind #6417.  Would you mind showing me the install result from #6303?

**Comment by sav-ix** [comment]
The `make install` output of https://github.com/openssl/openssl/pull/6303 :
```
	"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.14.26428\bin\HostX64\x64\nmake.exe" /                   depend && "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.14.26428\bin\HostX64\x64\nmake.exe" /                   _all
*** Installing development files
created directory `C:/libOPENSSL-1.1.1-dev/MSVC64RH'
created directory `C:/libOPENSSL-1.1.1-dev/MSVC64RH/include'
created directory `C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl'
Copying: ./include/openssl/aes.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/aes.h

<snip>

Copying: ./include/openssl/x509v3err.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/x509v3err.h
Copying: ./include/openssl/__DECC_INCLUDE_EPILOGUE.H to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/__DECC_INCLUDE_EPILOGUE.H
Copying: ./include/openssl/__DECC_INCLUDE_PROLOGUE.H to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/__DECC_INCLUDE_PROLOGUE.H
Copying: ./include/openssl/aes.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/aes.h
```

**Comment by levitte** [comment]
I just tested a build of master with the VC-WIN64A config target, and it doesn't install the `__DECC_*.h` files...

**Comment by sav-ix** [comment]
Odd. Can it be that we're using different shells for MSVC builds? (me Windows Command Prompt while you mintty or something)
The build command, which reproduces this issue on my system:
```
perl Configure --prefix="%CD%\..\MSVC64RH" --openssldir="%CD%\..\MSVC64RH\ssl" shared VC-WIN64A && nmake && nmake install

<snip>

	"C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.14.26428\bin\HostX64\x64\nmake.exe" /                   depend && "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.14.26428\bin\HostX64\x64\nmake.exe" /                   _all
*** Installing development files
created directory `C:/libOPENSSL-1.1.1-dev/MSVC64RH'
created directory `C:/libOPENSSL-1.1.1-dev/MSVC64RH/include'
created directory `C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl'
Copying: ./include/openssl/aes.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/aes.h
Copying: ./include/openssl/asn1.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/asn1.h

<snip>

Copying: ./include/openssl/x509v3err.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/x509v3err.h
Copying: ./include/openssl/__DECC_INCLUDE_EPILOGUE.H to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/__DECC_INCLUDE_EPILOGUE.H
Copying: ./include/openssl/__DECC_INCLUDE_PROLOGUE.H to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/__DECC_INCLUDE_PROLOGUE.H
Copying: ./include/openssl/aes.h to C:/libOPENSSL-1.1.1-dev/MSVC64RH/include/openssl/aes.h
```

**Comment by levitte** [comment]
`util/copy.pl` is a perl script, so question is why it's not running right. Did you run from a fresh master checkout?

**Comment by sav-ix** [comment]
Using 6da34cfbddede5e46f9c9183b724c99999dcfb41 master. And [ActivePerl x64 5.26.1.2601](https://www.activestate.com/activeperl/downloads), in case that matters.
Can anyone else test builds using MSVC under *native* Windows environment? If my issue not reproducible, there is no sense to waste time on it.

**Comment by levitte** [comment]
I did build under the same kind of conditions, the only difference being that I use Strawberry perl...

**Comment by sav-ix** [comment]
That could be a source of difference in our builds.
BTW, [OpenSSL Documentation](https://github.com/openssl/openssl/blob/master/NOTES.PERL)
>recommend ActiveState Perl

for MSVC builds.

**Comment by levitte** [comment]
I very much doubt that...  Appveyor uses ActiveState perl, as far as I know, and from the log of their [most recent build of master](https://ci.appveyor.com/project/openssl/openssl/build/master.18354/job/0ob8e0eb5e9ddy4f), it doesn't seem to make a difference.

**Comment by levitte** [comment]
Ok, going through the last uncheck ones...

**Comment by levitte** [comment]
9. Install `mingw-w64-x86_64-zlib` (or `mingw-w64-i686-zlib` for the 32-bit case).  If you did install `mingw-w64-x86_64-gcc` (see latest NOTES.WIN), you should get this automatically.

**Comment by levitte** [comment]
6.  I cannot reproduce with current master

**Comment by levitte** [comment]
7.  I'm willing to install tsget.pl and make tsget a link or copy to it on mingw, for the sake of backward compatibility.

**Comment by levitte** [comment]
7 is fixed by #6764

**Comment by levitte** [comment]
#6764 made it into master.  I think that finalizes all issues collected here.

For the future, I would rather have one issue per item instead of a collection of items in one issue, @sav-ix.  Makes it a lot easier to have a reasonable issue <-> fix connection.

**Comment by sav-ix** [comment]
> #6764 made it into master. I think that finalizes all issues collected here.

Yes. Thanks to all OpenSSL Developers, who helped with this issue.
